src/test/fio/fio_ceph_objectstore.cc

```diff
@@ -0,0 +1,393 @@
+
+#include <fio.h>
+#include <optgroup.h>
+
+#include "include/assert.h" // fio.h clobbers our assert.h
+
+
+// enable boost::intrusive_ptr<CephContext>
+void intrusive_ptr_add_ref(CephContext* cct) { cct->get(); }
+void intrusive_ptr_release(CephContext* cct) { cct->put(); }
+
+namespace {
+
+/// fio configuration options read from the job file
+struct Options {
+  thread_data* td;
+  char* conf;
+};
+
+template <class Func> // void Func(fio_option&)
+fio_option make_option(Func&& func)
+{
+  // zero-initialize and set common defaults
+  auto o = fio_option{};
+  o.category = FIO_OPT_C_ENGINE;
+  o.group    = FIO_OPT_G_RBD;
+  func(std::ref(o));
+  return o;
+}
+
+static std::vector<fio_option> ceph_options{
+  make_option([] (fio_option& o) {
+    o.name   = "conf";
+    o.lname  = "ceph configuration file";
+    o.type   = FIO_OPT_STR_STORE;
+    o.help   = "Path to a ceph configuration file";
+    o.off1   = offsetof(Options, conf);
+  }),
+  {} // fio expects a 'null'-terminated list
+};
+
+
+/// global engine state shared between all jobs within the process. this
+/// includes g_ceph_context and the ObjectStore instance
+struct Engine {
+  /// the initial g_ceph_context reference to be dropped on destruction
+  boost::intrusive_ptr<CephContext> cct;
+  std::unique_ptr<ObjectStore> os;
+
+  Engine(const thread_data* td);
+  ~Engine();
+
+  static Engine* get_instance(thread_data* td) {
+    // note: creates an Engine with the options associated with the first job
+    static Engine engine(td);
+    return &engine;
+  }
+};
+
+Engine::Engine(const thread_data* td)
+{
+  // add the ceph command line arguments
+  auto o = static_cast<const Options*>(td->eo);
+  if (!o->conf) {
+    throw std::runtime_error("missing conf option for ceph configuration file");
+  }
+  std::vector<const char*> args{
+    "-i", "0", // identify as osd.0 for osd_data and osd_journal
+    "--conf", o->conf, // use the requested conf file
+  };
+  if (td->o.directory) { // allow conf files to use ${fio_dir} for data
+    args.emplace_back("--fio_dir");
+    args.emplace_back(td->o.directory);
+  }
+
+  global_init(nullptr, args, CEPH_ENTITY_TYPE_OSD, CODE_ENVIRONMENT_UTILITY, 0);
+  common_init_finish(g_ceph_context);
+
+  // claim the g_ceph_context reference and release it on destruction
+  cct = boost::intrusive_ptr<CephContext>(g_ceph_context, false);
+
+  // create the ObjectStore
+  os.reset(ObjectStore::create(g_ceph_context,
+                               g_conf->osd_objectstore,
+                               g_conf->osd_data,
+                               g_conf->osd_journal));
+  if (!os)
+    throw std::runtime_error("bad objectstore type " + g_conf->osd_objectstore);
+
+  int r = os->mkfs();
+  if (r < 0)
+    throw std::system_error(-r, std::system_category(), "mkfs failed");
+
+  r = os->mount();
+  if (r < 0)
+    throw std::system_error(-r, std::system_category(), "mount failed");
+}
+
+Engine::~Engine()
+{
+  os->umount();
```

Comment:
thanks, that could indeed cause problems in the future. i'll need to revisit which callbacks fio provides, but i don't think it has a global init/shutdown for the mount and unmount, and that's why i had to go with Engine as a static